Inline variance operators for generic type variables: ``+T`` and ``-T`` (#2045)
python/peps: pep-9999.rst

parameter list of ``typing.Generic`` or ``typing.Protocol`` base classes,
and only there. It is not allowed to be used on type variables that
are already marked as covariant or contravariant, either with 
``covariant=True``, ``contravariant=True`` or with the new ``+T`` or 
``-T`` syntax.

Simplified Syntax
-----------------

Instead of specifying the variance on the typevar itself::

    T_co = typing.TypeVar('T_co', covariant=True)
    T_contra = typing.TypeVar('T_contra', contravariant=True)

    class SupportsInvertOld(typing.Protocol[T_co]):
        def __invert__(self) -> T_co: ...

    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
        def __le__(self, other: T_contra): ...

The new syntax uses the ``+`` and ``-`` prefix operators to specify
covariant and contravariant generic types inline::

    T = typing.TypeVar('T')

    class SupportsInvert(typing.Protocol[+T]):
        def __invert__(self) -> T: ...

    class SupportsPartialOrder(typing.Protocol[-T]):
        def __le__(self, other: T): ...

Differences with current syntax
-------------------------------

The new typevar operators return a transparent wrapper around the 

> original type variable, which can be accessed with the ``__orig__``

Reviewer: Should this not be ``__origin__`` to match the other generic aliases for run time inspection?

Author: Ah, yes, good catch, that is what I was going for. I'll correct it.

<https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance>`_
of `user-defined generic types
<https://www.python.org/dev/peps/pep-0484/#user-defined-generic-types>`_::

    T_co = typing.TypeVar('T_co', covariant=True)
    T_contra = typing.TypeVar('T_contra', contravariant=True)

    # a covariant generic type
    class EventDispatcher(typing.Generic[T_co]):
        def dispatch() -> T_co: ...

    # a contravariant generic type
    class EventListener(typing.Generic[T_contra]):
        def receive(event: T_contra) -> None: ...

This PEP proposes a concise syntax for specifying the variance of
generic types by overloading the ``+`` and ``-`` prefix operators
on type variables::

    T = typing.TypeVar('T')

    # a covariant generic type
    class EventDispatcher(typing.Generic[+T]):
        def dispatch() -> T: ...

    # a contravariant generic type
    class EventListener(typing.Generic[-T]):
        def receive(event: T) -> None: ...

Rationale
=========

Although defining the variance of generic type by using type variables
that are constructed with the ``covariant=True`` or ``contravariant=True``

> keyword arguments, it has downsides. 

Reviewer 2: Incomplete sentence

> This is why type checkers do no allow using co- or contravariant type

Reviewer 2: do *not*.

Also, it's not clear what the "This is why" refers to.

variables outside of generic classes. 

- In PEP 484 it states:

    Covariance or contravariance is not a property of a type variable,
    but a property of a generic class defined using this variable. 

  Instead of defining variance on the type variable itself, it would

>   make more sense to mark the type variable as variant when defining

Reviewer 2: ```suggestion
  make more sense to mark the type variable's variance when defining
```

  the generic type, and only there. 

- Because variance is a property of the generic type itself, type 

>   checkers do no allow using co- or contravariant type variables 

Reviewer 2: ```suggestion
  checkers do not allow using co- or contravariant type variables 
```

  outside of generic classes. But as the variance is currently a

>   property of the type variables, the must be used within variant

Reviewer 2: ```suggestion
  property of the type variables, they must be used within
```

  generic classes.

> - Creating separate type variables for different variances often

Reviewer 2: Not sure this is really common. It would be useful to have examples from open source projects where you see this duplication.

are already marked as covariant or contravariant, either with 
`covariant=True`, `contravariant=True` or with the new `+T` or `-T`
syntax.

Simplified Syntax
-----------------

Instead of specifying the variance on the typevar itself::

    T_co = typing.TypeVar('T_co', covariant=True)
    T_contra = typing.TypeVar('T_contra', contravariant=True)

    class SupportsInvertOld(typing.Protocol[T_co]):
        def __invert__(self) -> T_co: ...

    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
        def __le__(self, other: T_contra): ...

The new syntax uses the `+` and `-` prefix operators to specify
covariant and contravariant generic types inline::

    T = typing.TypeVar('T')

    class SupportsInvert(typing.Protocol[+T]):
        def __invert__(self) -> T: ...

    class SupportsPartialOrder(typing.Protocol[-T]):
        def __le__(self, other: T): ...

Differences with current syntax
-------------------------------

The new typevar operators return a transparent wrapper around the 
original type variable, which can be accessed with the `__orig__` 
attribute on the returned wrapper. e.g.::

    (+T).__orig__ is T

>     (+T).__covariant__ == not T.__covariant

Reviewer 2: ```suggestion
    (+T).__covariant__ == not T.__covariant__
```
Although why not just say `(+T).__covariant__ is True`?

    (+T).__contravariant__ is False
    (+T).__name__ == T.__name__
    (+T).__constraints__ == T.__constraints__
    (+T).__bound__ is T.__bound__

Thus, type variables defined with `covariant=True` and 
`contravariant=True`, are not equivalent to `+T` and `-T`.

`+T` and `-T` are not valid type annotations, and should only be used
within the generic type parameter list of generic classes. e.g.:: 

    class Spam(typing.Generic[+KT]): ...
    class Eggs(typing.Protocol[KT, +VT]): ...

>     class OrderedSet(typing.AbstractSet[+T]): ...

Reviewer 2: This is invalid according to the specification, which says you can only put it in Generic or Protocol. I was actually going to ask whether this should be allowed.

Author: Ah that's good to know. I can see how relaxing that restriction can be desirable, but I guess that that's out of scope here.

Reviewer 2: Is it a relaxing? This syntax is definitely legal right now without the `+`. If you allow it with your new syntax, you'll have to decide what it means. For example, maybe it should be possible to have an invariant subclass of a covariant class.

Reviewer 2: Actually, thinking about it a little more I believe you should allow this, because invariant subclasses of covariant classes are commonplace. My suggestion would be that you have to specify the `+` again for every subclass, and if there's no `+` the subclass is invariant.

For example, we have this in the standard library:
```
class AbstractSet(Generic[+T]): ...  # covariant

class set(Sequence[T]): ...  # invariant

class frozenset(Sequence[+T]): ... # covariant
```
To express that, we need what I suggested above.

Reviewer: I also agree with Jelle here, I think it should be allowed anywhere the the old forms are allowed.
On another note typing.AbstractSet is deprecated in favour of collections.abc.Set, so it might be a good idea to change that.

Reviewer 2: Another edge case is this:
```
class WhatIsMyVariance(Generic[+T], Sequence[-T]): ...
```
That should probably be disallowed, but this should be spelled out.

Author: I agree with your suggestion @JelleZijlstra . But I think that In your example `set` and `frozenset` also need to subclass `Generic`, i.e.

```python

class set(Sequence[T], Generic[T]): ...  # invariant

class frozenset(Sequence[T], Generic[+T]): ... # covariant
```

Then the restriction that `+T` and `-T` can only be used within the type parameter list of `Generic` and `Protocol` still holds. If you were to instead write `class frozenset(Sequence[+T], Generic[+T]): ...`, to me it looks as if `Sequence` itself is marked as covariant under `T`, whereas the covariance only applies to the specific definition of `frozenset`; the variance information does not need to be passed to the `Sequence` superclass. 
This also avoids the possibility for edge cases, like in `WhatIsMyVariance`.

Reviewer 2: That's not how things currently work, though: writing `class set(Sequence[T]):` is legal and accepted by typecheckers.

Author: Ah I see, somehow I thought that in that case you couldn't do e.g. `spam: set[str]`, and that it required an explicit `Generic` as additional superclass. 

Anyway, I now indeed see how this behavior has to be explicitly defined. Requiring the same variance for the same type variable in the same generic class should do the trick. So no mixing; only allow e.g.
- `class LinkedList(Iterable[T], Container[T]): ... `
- `class LinkedList(Iterable[+T], Container[+T]): ... ` 
- `class LinkedList(Iterable[-T], Container[-T]): ... ` 

This could be checked within `typing.Generic.__init_subclass__`.
